stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+                if let Err(e) = self
+                    .stackerdb
+                    .send_message_with_retry(block_rejection.into())
+                {
+                    warn!(
+                        "Signer #{}: Failed to send block rejection to stacker-db: {e:?}",
+                        self.signer_id
+                    );
+                }
+            }
+            is_valid
+        } else {
+            // Failed to connect to the stacks node to get transactions. Cannot validate the block. Reject it.
+            debug!(
+                "Signer #{}: Broadcasting a block rejection due to signer connectivity issues...",
+                self.signer_id
+            );
+            let block_rejection = BlockRejection::new(
+                block.header.signer_signature_hash(),
+                RejectCode::ConnectivityIssues,
+            );
+            // Submit signature result to miners to observe
+            if let Err(e) = self
+                .stackerdb
+                .send_message_with_retry(block_rejection.into())
+            {
+                warn!(
+                    "Signer #{}: Failed to send block submission to stacker-db: {e:?}",
+                    self.signer_id
+                );
+            }
+            false
+        }
+    }
+
+    /// Filter out transactions from the stackerdb that are not valid
+    /// i.e. not valid vote-for-aggregate-public-key transactions from registered signers
+    fn filter_invalid_transactions(
+        &self,
+        stacks_client: &StacksClient,
+        current_reward_cycle: u64,
+        signer_slot_ids: &HashMap<StacksAddress, u32>,
+        transaction: StacksTransaction,
+    ) -> Option<StacksTransaction> {
+        // Filter out transactions that have already been confirmed (can happen if a signer did not update stacker db since the last block was processed)
+        let origin_address = transaction.origin_address();
+        let origin_nonce = transaction.get_origin_nonce();
+        let Some(origin_signer_id) = signer_slot_ids.get(&origin_address) else {
+            debug!(
+                "Signer #{}: Unrecognized origin address ({origin_address}). Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        let Ok(account_nonce) = retry_with_exponential_backoff(|| {
+            stacks_client
+                .get_account_nonce(&origin_address)
+                .map_err(backoff::Error::transient)
+        }) else {
+            warn!(
+                "Signer #{}: Unable to get account for transaction origin address: {origin_address}. Filtering ({}).",
+                self.signer_id,
+                transaction.txid()
+            );
+            return None;
+        };
+        // TODO: add a check that we don't have two conflicting transactions in the same block from the same signer. This is a potential attack vector (will result in an invalid block)
```

Comment:
An alternative rule that's easier to enforce would be to require that each signer is entitled to at most one stackerdb-submitted transaction per block.  We'd only ever try to decode a single transaction from the signer's stackerdb slot, and we'd ignore it if it was already mined (e.g. by checking the signer's nonce against the tx nonce).

Reply:
Yeah I have a comment in the miners section with this potential solution ^ I will do as part of my transaction revamp!